src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorEditorFactory.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            if (!IsLSPRazorEditorEnabled)
+            {
+                docView = default;
+                docData = default;
+                editorCaption = null;
+                cmdUI = default;
+                cancelled = 0;
+
+                // Razor LSP is not enabled, allow another editor to handle this document
+                return VSConstants.VS_E_UNSUPPORTEDFORMAT;
+            }
+
+            var editorInstance = base.CreateEditorInstance(createDocFlags, moniker, physicalView, pHier, itemid, existingDocData, out docView, out docData, out editorCaption, out cmdUI, out cancelled);
+            var textLines = (IVsTextLines)Marshal.GetObjectForIUnknown(docData);
+
+            SetTextBufferContentType(textLines);
+
+            return editorInstance;
+        }
+
+        private void SetTextBufferContentType(IVsTextLines textLines)
+        {
+            var textBufferDataEventsGuid = typeof(IVsTextBufferDataEvents).GUID;
+            var connectionPointContainer = textLines as IConnectionPointContainer;
+            connectionPointContainer.FindConnectionPoint(textBufferDataEventsGuid, out var connectionPoint);
+            var contentTypeSetter = new TextBufferContentTypeSetter(
+                textLines,
+                _adaptersFactory.Value,
+                _razorLSPContentType.Value);
+            contentTypeSetter.Attach(connectionPoint);
+
+            // Next, the editor typically resets the ContentType after TextBuffer creation. We need to let them know
+            // to not update the content type.
+            var userData = textLines as IVsUserData;
+            var hresult = userData.SetData(VSConstants.VsTextBufferUserDataGuid.VsBufferDetectLangSID_guid, false);
+
+            ErrorHandler.ThrowOnFailure(hresult);
+        }
+
+        private class TextBufferContentTypeSetter : IVsTextBufferDataEvents
+        {
+            private readonly IVsTextLines _textLines;
+            private readonly IVsEditorAdaptersFactoryService _adaptersFactory;
+            private readonly IContentType _razorLSPContentType;
+            private IConnectionPoint _connectionPoint;
+            private uint _connectionPointCookie;
+
+            public TextBufferContentTypeSetter(
+                IVsTextLines textLines,
+                IVsEditorAdaptersFactoryService adaptersFactory,
+                IContentType razorLSPContentType)
+            {
+                _textLines = textLines;
+                _adaptersFactory = adaptersFactory;
+                _razorLSPContentType = razorLSPContentType;
+            }
+
+            public void Attach(IConnectionPoint connectionPoint)
+            {
+                if (connectionPoint is null)
+                {
+                    throw new ArgumentNullException(nameof(connectionPoint));
+                }
+
+                _connectionPoint = connectionPoint;
+
+                connectionPoint.Advise(this, out _connectionPointCookie);
```

Comment:
similarly - does this have an error code?

Reply:
Nope, void returning and no other overloads